### PR TITLE
Make Sapper server more self-contained

### PR DESCRIPTION
**Pull request by Rich-Harris**
This PR may not be totally uncontroversial, so I'd welcome feedback on it.

In #660 I proposed incorporating `sirv` directly into Sapper — we're already replicating some of its functionality (to serve client assets), and the `static` folder is 'special' to a Sapper project, so it kind of makes sense. And because it's a @lukeed project, it's small enough that it's basically free.

Right now, the template `src/server.js` looks like this:

```js
import sirv from 'sirv';
import polka from 'polka';
import compression from 'compression';
import * as sapper from '@sapper/server';

const { PORT, NODE_ENV } = process.env;
const dev = NODE_ENV === 'development';

polka() // You can also use Express
	.use(
		compression({ threshold: 0 }),
		sirv('static', { dev }),
		sapper.middleware()
	)
	.listen(PORT, err => {
		if (err) console.log('error', err);
	});
```

Arguably, `compression` doesn't belong at this layer (it should be handled by your CDN), so we could simplify that to this..

```diff
-import sirv from 'sirv';
import polka from 'polka';
-import compression from 'compression';
import * as sapper from '@sapper/server';

-const { PORT, NODE_ENV } = process.env;
-const dev = NODE_ENV === 'development';
+const { PORT } = process.env;

polka() // You can also use Express
	.use(
-		compression({ threshold: 0 }),
-		sirv('static', { dev }),
		sapper.middleware()
	)
	.listen(PORT, err => {
		if (err) console.log('error', err);
	});
```

But then there's really no need for Polka — we can just do this:

```diff
-import polka from 'polka';
+import * as http from 'http';
import * as sapper from '@sapper/server';

const { PORT } = process.env;

-polka() // You can also use Express
-	.use(
-		sapper.middleware()
-	)
+http.createServer(sapper.middleware())
	.listen(PORT, err => {
		if (err) console.log('error', err);
	});
```

We can make that even nicer by exporting a `start` function:

```js
import * as sapper from '@sapper/server';

sapper.start();
```

(This PR also adds a `sapper.server(opts)` function, which just creates the server but doesn't listen on `process.env.PORT`.)

This also gives us a nice symmetry with `src/client.js` (especially with #645). **The trade-off is** that the barrier to understanding how to add your own custom non-Sapper routes is marginally higher, which does run counter to Sapper's 'show the guts' philosophy.

Not shown in this PR — a `sapper.render(...)` or `sapper.handle(...)` function. I think we should add that, probably in a separate PR, I'm just not 100% sure what the API should be yet.

Thoughts?

### CODE REVIEW

File: runtime/src/server/middleware/index.ts [hunk 1]
```diff
@@ -1,23 +1,42 @@
...
+
+		sirv(`${build_dir}/client`, {
+			dev,
+			maxAge: 31536000,
+			immutable: true
+		}),
+
+		sirv(`${build_dir}/service-worker`, {
+			dev,
+			maxAge: 300
```
Comment (lukeed) [inline runtime/src/server/middleware/index.ts]: Did you mean to change the SW settings? Was no-cache, must-revalidate, etc

**Comment by ansarizafar** [comment]
I would suggest sapper.use() for using middlewares like helmet, bodyparser etc.

File: runtime/src/server/middleware/index.ts (hunk 1 above)
Comment (Rich-Harris) [reply runtime/src/server/middleware/index.ts]: huh, I guess not. Not sure what I was thinking. How would I add those with sirv?

Comment (lukeed) [reply runtime/src/server/middleware/index.ts]: It's custom settings, so has to be thru `setHeaders` option.

**Comment by lukeed** [comment]
Perhaps I missed it, but I think sirv needs to be moved to dependencies list

**Comment by mrkishi** [comment]
I like it!

Do you mean `sapper.handle()` would expose sapper as a middleware? Because while `sapper.start()` is awesome for the main api, removing that option seems too closed-warden. If you meant it as a way to add custom routes, I think at this point `sapper.start()` could just return a sapper server with an usual routing api (`get`, `post`, etc), since it's basically acting as a server framework.

I'd really love to go for the direction where sapper's server provides a better internal api, and just have the main interface (`sapper.start`) use said internal api. We could then expose the internal api for advanced use-cases. I haven't had the chance to give it proper thought, but the holy grail would be granular enough to be able to expose treeshakeable rendering at the route level from `@sapper/server` (think lambda per route, routing info available to be converted to whatever lambda provider you use).

File: runtime/src/server/middleware/index.ts [hunk 2]
```diff
@@ -1,23 +1,42 @@
...
+		}),
+
+		function condition_request(req: Req, res: Res, next: () => void) {
+			const qi = req.url.indexOf('?');
+			req.query = ~qi ? querystring.parse(req.url.slice(qi + 1)) : {}
+
 			if (req.baseUrl === undefined) {
-				let { originalUrl } = req;
+				let originalUrl = req.originalUrl || req.url;
```
Comment (benmccann) [inline runtime/src/server/middleware/index.ts:39]: This change is also being requested in https://github.com/sveltejs/sapper/pull/942

**Pull request closed without merging**